src/site/artifacts/mod.rs

```diff
@@ -1,76 +1,143 @@
 use std::collections::HashMap;
 
 use crate::config::Config;
-use crate::data::artifacts::{DisplayPreference, InstallMethod};
+use crate::data::artifacts::{File, FileIdx, InstallMethod, InstallerIdx, TargetTriple};
 use crate::data::{Context, Release};
 use crate::errors::*;
 
-mod installers;
-mod table;
+use crate::data::artifacts::inference::triple_to_display_name;
+use crate::message::{Message, MessageType};
+use crate::site::javascript;
+use serde::Serialize;
 
-use axohtml::elements::div;
-use axohtml::{html, text};
+type DownloadableFiles = Vec<(FileIdx, (File, Vec<Option<String>>))>;
+type Platforms = HashMap<TargetTriple, Vec<InstallerIdx>>;
 
-pub fn page(context: &Context, config: &Config) -> Result<String> {
+#[derive(Serialize, Debug)]
+pub struct ArtifactsContext {
+    tag: String,
+    formatted_date: Option<String>,
+    platforms_with_downloads: Platforms,
+    downloadable_files: DownloadableFiles,
+    release: Release,
+    os_script: String,
+    has_checksum_files: bool,
+}
+
+pub fn template_context(context: &Context, config: &Config) -> Result<Option<ArtifactsContext>> {
     let Some(release) = context.latest() else {
-        return Ok(String::new());
+        return Ok(None);
     };
+    let os_script = javascript::build_os_script_path(&config.build.path_prefix);
+    let platforms_with_downloads = filter_platforms(release);
+
+    let mut downloadable_files: Vec<_> = release
+        .artifacts
+        .installers()
+        .filter_map(|(_, installer)| {
+            let InstallMethod::Download { file } = installer.method else {
+            return None;
+        };
```

Comment:
wait did rustfmt actually generate this? 0.o

(i think rustfmt support for let-else might only be on the latest toolchain)